java/main/jglome/Glome.java

```diff
@@ -0,0 +1,357 @@
+  }
+
+  public XECPrivateKey userPrivateKey() {
+    return (XECPrivateKey) userKeys.getPrivate();
+  }
+
+  /**
+   * Generates a user's tag corresponding to the given message {@code msg} and the counter {@code
+   * cnt}.
+   *
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the user to the peer.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   */
+  public byte[] generateTag(byte[] msg, int cnt) throws CounterOutOfBoundsException {
+    return generateTag(msg, cnt, this.userMacKey);
+  }
+
+  /**
+   * Checks whether the peer's tag matches received message and some counter.
+   *
+   * @param peerTag tag from a peer.
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the peer to the user.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   * @throws WrongTagException if the peer's tag has invalid length (less than {@code
+   * minPeerTagLength} or more than {@code MAX_TAG_LENGTH}) or it's not equal to the prefix of a
+   * correct tag.
+   */
+  public void checkTag(byte[] peerTag, byte[] msg, int cnt)
+      throws CounterOutOfBoundsException, WrongTagException {
+    if (peerTag.length < minPeerTagLength || peerTag.length > MAX_TAG_LENGTH) {
+      throw new WrongTagException(
+          "The received tag has invalid length. Expected " + minPeerTagLength + " or more, got "
+              + peerTag.length + ".");
+    }
+
+    byte[] truncatedTag = Arrays.copyOf(generateTag(msg, cnt, this.peerMacKey), peerTag.length);
+    if (!Arrays.equals(peerTag, truncatedTag)) {
+      throw new WrongTagException("The received tag doesn't match the expected tag.");
+    }
+  }
+
+  /**
+   * Common part of the constructors. Initialises peer's public key, minimum length of a peer's tag,
+   * base point for Curve25519 curve. Calculates user's and peer's MAC keys.
+   *
+   * @param peerKey peer's public key.
+   * @param minPeerTagLength minimum length of a peer's tag.
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  private void initMacKeys(XECPublicKey peerKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException {
+    this.peerKey = peerKey;
+    if (minPeerTagLength < MIN_TAG_LENGTH || minPeerTagLength > MAX_TAG_LENGTH) {
+      throw new MinPeerTagLengthOutOfBoundsException(
+          "minPeerTagLength argument should be in [1..32] range. Got " + minPeerTagLength + ".");
+    }
+    this.minPeerTagLength = minPeerTagLength;
+
+    KeyAgreement ka;
+    try {
+      ka = KeyAgreement.getInstance("XDH");
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError(e.getMessage());
+    }
+
+    try {
+      ka.init(userPrivateKey());
+      ka.doPhase(peerKey, true);
+    } catch (InvalidKeyException e) {
+      throw new AssertionError(e.getMessage());
```

Comment:
When does this happen and is this something that can happen on invalid input? In that case (contrary to the system exception above), I'd rather we expose the failure to the caller as an Exception.

Reply:
This Exception is caused both on .init() and .doPhase() methods and yes, happens on invalid input keys. Thank you for your idea, sounds good.